### DISCUSSION TITLE
Question about PersistedQueryCache deprecation and replacement interface

### DISCUSSION THREAD

**Discussion by softprops** [graphql-java/graphql-java#3342] [Q&A] [answered]
`getPersistedQueryDocument` is now deprecated in favor of `getPersistedQueryDocumentAsync` which returns a `CompletedFuture` but I'm curious about [the new api](https://github.com/graphql-java/graphql-java/blob/aae306959581bc92b2a92b525c61cf82ac869e83/src/main/java/graphql/execution/preparsed/persisted/PersistedQueryCache.java#L56) 

The method still declares `throws PersistedQueryNotFound` but presumably that would happen inside the asynchronous operation at which point an implementation with throw `PersistedQueryNotFound` inside the async block but that wouldn't actually bubble up as a thrown exception according to the interface it would result in a failed `CompletableFuture`.

Can someone help me reason about how implementors should implement the new api?

**Comment by bbakerman** [comment, marked as answer]
Because we used defaulted interface methods here, the older `getPersistedQueryDocument` is called like this

```java
default CompletableFuture<PreparsedDocumentEntry> getPersistedQueryDocumentAsync(Object persistedQueryId, ExecutionInput executionInput, PersistedQueryCacheMiss onCacheMiss) throws PersistedQueryNotFound{
        return CompletableFuture.completedFuture(getPersistedQueryDocument(persistedQueryId, executionInput, onCacheMiss));
    }
```    

So the default method is still able to throw `PersistedQueryNotFound`.

If you write your own implmentation using a `CompletableFuture` then by all means return any exceptions back via the CF.

Any exceptions thrown will be propagated back up to the top of the graphql `graphql.GraphQL#executeAsync(graphql.ExecutionInput)`

**Reply by softprops** [reply]
> return any exceptions back via the CF

cool I think that's the context I was missing.  I wasn't sure if `CompletableFuture.failedFuture(new PersistedQueryNotFound(persistedQueryId))` would be treated the same.

thanks @bbakerman
